### PR TITLE
`[Fix]` Part-2 - LevelDbStore

### PR DESCRIPTION
**Pull request by cschuchardt88**
# Description

This fixes the store's `dotnet` memory management for `GC`, a long with adding new feature options, easier `Exception` catching. 

# Change Log
 - Added `LevelDbStoreTests` for unit testing `DB` and `Iterator`

## Type of change

- [x] Optimization (the change is only an optimization)
- [x] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Unit tests

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/Plugins/LevelDBStore/Plugins/Storage/Store.cs [hunk 1]
```diff
@@ -9,59 +9,65 @@
...
+        public Store(string dir)
         {
-            db = DB.Open(path, new Options { CreateIfMissing = true, FilterPolicy = Native.leveldb_filterpolicy_create_bloom(15) });
+            _db = new(dir, new()
+            {
+                CreateIfMissing = true,
+                // Keep whole blockchain open plus future files
+                // at lease up to block index 10_000_000
+                MaxOpenFiles = 4096,
+                FilterPolicy = 10,
```
Comment (shargon) [inline src/Plugins/LevelDBStore/Plugins/Storage/Store.cs:31]:
```suggestion
                FilterPolicy = Native.leveldb_filterpolicy_create_bloom(15),
```

?

File: src/Plugins/LevelDBStore/Plugins/Storage/Snapshot.cs [hunk 2]
```diff
@@ -9,61 +9,48 @@
...
+            _db.Get(key, _readOptions);
 
-        public byte[] TryGet(byte[] key)
-        {
-            return db.Get(options, key);
-        }
+        public IEnumerable<(byte[] Key, byte[] Value)> Seek(byte[] prefix, SeekDirection direction = SeekDirection.Forward) =>
+            direction == SeekDirection.Forward
+                ? _db.Seek(prefix, new ReadOptions { FillCache = false, })
+                : _db.SeekPrev(prefix, new ReadOptions { FillCache = false, });
```
Comment (shargon) [inline src/Plugins/LevelDBStore/Plugins/Storage/Snapshot.cs:54]:
```suggestion
                ? _db.Seek(prefix, ReadOptions.Default)
                : _db.SeekPrev(prefix, ReadOptions.Default);
```

File: src/Plugins/LevelDBStore/IO/Data/LevelDB/WriteBatch.cs [hunk 3]
```diff
@@ -11,30 +11,66 @@
...
         {
-            Native.leveldb_writebatch_put(handle, key, (UIntPtr)key.Length, value, (UIntPtr)value.Length);
+            Native.leveldb_writebatch_iterate(Handle, state, put, deleted);
         }
+
+        protected override void FreeUnManagedObjects()
+        {
+            Native.leveldb_writebatch_destroy(Handle);
+        }
+
```
Comment (shargon) [inline src/Plugins/LevelDBStore/IO/Data/LevelDB/WriteBatch.cs:74]:
```suggestion
```

File: src/Plugins/LevelDBStore/IO/Data/LevelDB/LevelDBHandle.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+                {
+                    FreeManagedObjects();
+                }
+                if (Handle != nint.Zero)
+                {
+                    FreeUnManagedObjects();
+                    Handle = nint.Zero;
+                }
+                _disposed = true;
+            }
```
Comment (shargon) [inline src/Plugins/LevelDBStore/IO/Data/LevelDB/LevelDBHandle.cs:52]:
```suggestion
                _disposed = true;
                if (disposing)
                {
                    FreeManagedObjects();
                }
                if (Handle != nint.Zero)
                {
                    FreeUnManagedObjects();
                    Handle = nint.Zero;
                }
            }
```

File: src/Plugins/LevelDBStore/IO/Data/LevelDB/Native.cs [hunk 5]
```diff
@@ -10,71 +10,79 @@
...
+        NoCompression = 0,
+        SnappyCompression = 1,
     }
 
     public static class Native
     {
         #region Logger
         [DllImport("libleveldb", CharSet = CharSet.Ansi, CallingConvention = CallingConvention.Cdecl)]
-        public static extern IntPtr leveldb_logger_create(IntPtr /* Action<string> */ logger);
+        public static extern nint leveldb_logger_create(nint /* Action<string> */ logger);
```
Comment (shargon) [inline src/Plugins/LevelDBStore/IO/Data/LevelDB/Native.cs:27]: Is this change needed?

File: src/Plugins/LevelDBStore/Plugins/Storage/Store.cs (hunk 1 above)
Comment (cschuchardt88) [reply src/Plugins/LevelDBStore/Plugins/Storage/Store.cs:31]: No need for extra, 5 bytes. Bloom filters should be as small as possible for datasets without that many collisions. 10 is way more than enough.

File: src/Plugins/LevelDBStore/IO/Data/LevelDB/LevelDBHandle.cs (hunk 4 above)
Comment (cschuchardt88) [reply src/Plugins/LevelDBStore/IO/Data/LevelDB/LevelDBHandle.cs:52]: Leveldb `handle` isnt `disposed` until `FreeManagedObjects()` is completed and `Handle` is zeroed.

File: src/Plugins/LevelDBStore/Plugins/Storage/Snapshot.cs (hunk 2 above)
Comment (cschuchardt88) [reply src/Plugins/LevelDBStore/Plugins/Storage/Snapshot.cs]: This will cause problem, because `ReadOptions.Default` is `static readonly` and will use the same instance. So if someone changes `ReadOptions.Default` parameters will cause problems for us. Thats why I coded it this way.

File: src/Plugins/LevelDBStore/IO/Data/LevelDB/Native.cs (hunk 5 above)
Comment (cschuchardt88) [reply src/Plugins/LevelDBStore/IO/Data/LevelDB/Native.cs:27]: Yes, `nint` is a `dotnet` `keyword` and gets compiled in the `ilasm` that is much faster OpCode then using the `dotnet` `objects classes`

File: src/Plugins/LevelDBStore/Plugins/Storage/Snapshot.cs (hunk 2 above)
Comment (cschuchardt88) [reply src/Plugins/LevelDBStore/Plugins/Storage/Snapshot.cs]: `ReadOptions.Default` isnt the same as `new ReadOptions { FillCache = false, }`

**Comment by superboyiii** [comment]
@cschuchardt88 Need solve conflict.
